src/controller/tests/data_model/TestCommands.cpp

```diff
@@ -239,6 +247,53 @@ void TestCommandInteraction::TestSuccessNoDataResponse(nlTestSuite * apSuite, vo
     NL_TEST_ASSERT(apSuite, ctx.GetExchangeManager().GetNumActiveExchanges() == 0);
 }
 
+void TestCommandInteraction::TestAsyncResponse(nlTestSuite * apSuite, void * apContext)
+{
+    TestContext & ctx = *static_cast<TestContext *>(apContext);
+    TestCluster::Commands::TestSimpleArgumentRequest::Type request;
+    auto sessionHandle = ctx.GetSessionBobToAlice();
+
+    bool onSuccessWasCalled = false;
+    bool onFailureWasCalled = false;
+    bool statusCheck        = false;
+    request.arg1            = true;
+
+    // Passing of stack variables by reference is only safe because of synchronous completion of the interaction. Otherwise, it's
+    // not safe to do so.
+    auto onSuccessCb = [&onSuccessWasCalled, &statusCheck](const app::ConcreteCommandPath & commandPath,
+                                                           const app::StatusIB & aStatus, const auto & dataResponse) {
+        statusCheck        = (aStatus.mStatus == Protocols::InteractionModel::Status::Success);
+        onSuccessWasCalled = true;
+    };
+
+    // Passing of stack variables by reference is only safe because of synchronous completion of the interaction. Otherwise, it's
+    // not safe to do so.
+    auto onFailureCb = [&onFailureWasCalled](const app::StatusIB & aStatus, CHIP_ERROR aError) { onFailureWasCalled = true; };
+
+    responseDirective = kAsync;
+
+    chip::Controller::InvokeCommandRequest(&ctx.GetExchangeManager(), sessionHandle, kTestEndpointId, request, onSuccessCb,
+                                           onFailureCb);
+
+    NL_TEST_ASSERT(apSuite, !onSuccessWasCalled && !onFailureWasCalled && !statusCheck);
+    NL_TEST_ASSERT(apSuite, ctx.GetExchangeManager().GetNumActiveExchanges() == 1);
+
+    CommandHandler * commandHandle = asyncHandle.Get();
+    NL_TEST_ASSERT(apSuite, commandHandle != nullptr);
```

Comment:
How is this test passing? Without calls to `DrainAndServiceIO`, messages won't get dispatched..

Reply:
You'll need one call to that right after sending the command (to pump the message through on the RX side), and another call to that after you `AddStatus` below to get the incoming message back to the client processed).